### DISCUSSION TITLE
Deadlock detected when CALL partman.run_maintenance_proc() was invoked

### DISCUSSION THREAD

**Discussion by sriram-au** [pgpartman/pg_partman#555] [Q&A] [answered]
We started seeing a `deadlock detected` error in our production RDS postgres when run_maintenance_proc() was called by our `pg_cron` job that runs daily. We don't see this in our lower environments interestingly. Obviously our access patterns for prod DB is vastly different

Here is the exact error:
```
CALL partman.run_maintenance_proc();
ERROR:  deadlock detected
CONTEXT: SQL statement "ALTER TABLE public.command_record ATTACH PARTITION public.command_record_p2023w38 FOR VALUES FROM ('2023-09-18 00:00:00+00') TO ('2023-09-25 00:00:00+00')"
PL/pgSQL function create_partition_time(text,timestamp with time zone[],boolean,text) line 249 at EXECUTE
PL/pgSQL function partman.run_maintenance(text,boolean,boolean) line 284 at assignment
SQL statement "SELECT partman.run_maintenance('public.command_record', p_jobmon := 't')"
PL/pgSQL function partman.run_maintenance_proc(integer,boolean,boolean) line 42 at EXECUTE
DETAIL: Process 3198 waits for AccessExclusiveLock on relation 16755 of database 16404; blocked by process 3061.
Process 3061 waits for RowExclusiveLock on relation 16656 of database 16404; blocked by process 3198.
HINT: See server log for query details.
CONTEXT: PL/pgSQL function create_partition_time(text,timestamp with time zone[],boolean,text) line 511 at RAISE
PL/pgSQL function partman.run_maintenance(text,boolean,boolean) line 284 at assignment
SQL statement "SELECT partman.run_maintenance('public.command_record', p_jobmon := 't')"
PL/pgSQL function partman.run_maintenance_proc(integer,boolean,boolean) line 42 at EXECUTE
DETAIL:
HINT:
CONTEXT:  PL/pgSQL function partman.run_maintenance(text,boolean,boolean) line 413 at RAISE
SQL statement "SELECT partman.run_maintenance('public.command_record', p_jobmon := 't')"
PL/pgSQL function partman.run_maintenance_proc(integer,boolean,boolean) line 42 at EXECUTE
```

When I try to attach a partition manually there seems to be no problems.
```
CREATE TABLE command_record_p2023w38 PARTITION OF command_record FOR VALUES FROM ('2023-09-18 00:00:00+00') TO ('2023-09-25 00:00:00+00');
```

Any help is much appreciated.

**Comment by keithf4** [comment]
Only thing within pg_partman I could think of that could cause this is calling the maintenance procedure more than once at the same time.
So my first thought was that RDS might have the background worker enabled, interfering with calling it via pgcron. But reading their docs it doesn't seem like it does (https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/PostgreSQL_Partitions.html#PostgreSQL_Partitions.run_maintenance_proc). 

Only other thing I can think of is to double-check that you don't have multiple cronjobs for maintenance scheduled. 

If that's not the case, are you able to get the full deadlock log out of the PG logs that shows what all the actual queries for each process were doing at the time of the deadlock? You may have to adjust your log_line_prefix to ensure it includes the PID so you can match things up. If you are seeing two different PIDs both calling maintenance at the same time, then something else is calling maintenance at the same time as your cron.

**Reply by sriram-au** [reply]
Hey @keithf4 Thanks for the prompt response! We don't think there are multiple cron jobs. 
```
select * from cron.job;
 jobid | schedule |               command               | nodename  | nodeport |  database  | username  | active | jobname
-------+----------+-------------------------------------+-----------+----------+------------+-----------+--------+---------
     1 | @daily   | CALL partman.run_maintenance_proc() | localhost |     5432 | <redacted> | <redacted> | t      |
```

In the cron logs we only see 1 run every night. 😕 

We will look into getting the PG logs.

**Reply by keithf4** [reply]
Other thing to look into is what objects those relation numbers correspond to. Can either cast them to regclass (`SELECT 16755::regclass`) or look them up in the `pg_class` catalog. Should also help narrow down if something else my be touching the partitioned tables and interfering.

**Reply by sriram-au** [reply]
`16755` is our default partition and `16656` is our parent table. The parent table query is most likely from our application. Not all queries use the partitioning key. Hence, we see some queries resulting in sub transactions on all partitions. 
As for process that is trying to get an access exclusive lock on the default partition is most likely the `run_maintenance_proc` function. Is this statement correct? We are looking at the PG logs for more details.

**Reply by keithf4** [reply]
Do you have any data in your default?

**Reply by sriram-au** [reply]
Cron job failure error: (The one I pasted above was from an adhoc run of the `partman.run_maintenance_proc()` ):
```
ERROR: deadlock detected                                                                                                                                                           +
               | CONTEXT: SQL statement "ALTER TABLE public.command_record ATTACH PARTITION public.command_record_p2023w38 FOR VALUES FROM ('2023-09-18 00:00:00+00') TO ('2023-09-25 00:00:00+00')"+
               | PL/pgSQL function create_partition_time(text,timestamp with time zone[],boolean,text) line 249 at EXECUTE                                                                          +
               | PL/pgSQL function partman.run_maintenance(text,boolean,boolean) line 284 at assignment                                                                                             +
               | SQL statement "SELECT partman.run_maintenance('public.command_record', p_jobmon := 't')"                                                                                           +
               | PL/pgSQL function partman.run_maintenance_proc(integer,boolean,boolean) line 42 at EXECUTE                                                                                         +
               | DETAIL: Process 75794 waits for AccessExclusiveLock on relation 16755 of database 16404; blocked by process 75511.                                                                 +
               | Process 75511 waits for RowExclusiveLock on relation 16656 of database 16404; blocked by process 75794.                                                                            +
               | HINT: See server log for query details.                                                                                                                                            +
               | CONTEXT: PL/pgSQL function create_partition_time(text,timestamp with time zone[],boolean,text) line 511 at RAISE                                                                   +
               | PL/pgSQL function partman.run_maintenance(text,boolean,boolean) line 284 at assignment                                                                                             +
               | SQL statement "SELECT partman.run_maintenance('public.command_record', p_jobmon := 't')"                                                                                           +
               | PL/pgSQL function partman.run_maintenance_proc(integer,boolean,boolean) line 42 at EXECUTE                                                                                         +
               | DETAIL:                                                                                                                                                                            +
               | HINT:                                                                                                                                                                              +
               | CONTEXT: PL/pgSQL function partman.run_maintenance(text,boolean,boolean) line 413 at RAISE                                                                                         +
               | SQL statement "SELECT partman.run_maintenance('public.command_record', p_jobmon := 't')"                                                                                           +
```

Looking at the logs, the deadlock is caused by the maintenance process `75794` is waiting for an AccessExclusiveLock on relation `16755` (which is our default partition) and  process `75511` which is an update query performed by our application waiting for a  RowExclusiveLock on relation `16656`.  (Which is our parent table)

`75794` seems to be holding the lock that `75511` is looking for 

```
692316801000,2023-08-18 00:00:01 UTC:172.21.3.176(42887):readwrite@commandapi:[75528]:LOG:  process 75528 still waiting for RowExclusiveLock on relation 16656 of database 16404 after 1000.037 ms at character 13
1692316801000,"2023-08-18 00:00:01 UTC:172.21.3.176(42887):readwrite@commandapi:[75528]:DETAIL:  Processes holding the lock: 75794, 75474, 75545, 75504, 75629, 75604, 75608, 75449, 75423, 75425, 75498, 75459, 75617, 75489, 75442, 75483, 75680, 75539, 75426, 75647, 75446, 75418, 75534, 75436, 75473, 75598, 75477, 75417, 75682, 75427, 75655, 75499, 75527, 75553, 75666, 75430, 75424, 75413, 75615, 75420, 75421, 75454, 75428, 75525, 75652, 75599, 75500, 75605, 75437, 75448, 75501, 75431, 75429, 75435, 75419, 75460, 75669, 75445, 75438, 75444, 75585, 75455, 75432, 75673, 75434, 75440, 75519, 75461, 75443, 75439, 75466, 75531, 75450, 75638, 75433, 75441, 75451, 75465, 75536, 75447, 75486, 75452, 75540, 75490, 75422, 75464, 75532. Wait queue: 75511, 75528."
```

Similarly `75511` is holding that lock that `75794` is looking for

```
 process 75794 detected deadlock while waiting for AccessExclusiveLock on relation 16755 of database 16404 after 1000.280 ms
1692316801000,"2023-08-18 00:00:01 UTC::@:[75794]:DETAIL:  Processes holding the lock: 75794, 75511, 75528, 75474, 75545, 75504, 75629, 75604, 75608, 75449, 75423, 75425, 75498, 75459, 75617, 75489, 75442, 75483, 75680, 75539, 75426, 75446, 75647, 75418, 75534, 75436, 75473, 75598, 75477, 75417, 75682, 75427, 75655, 75499, 75527, 75553, 75666, 75430, 75424, 75413, 75615, 75420, 75421, 75454, 75428, 75525, 75652, 75599, 75500, 75605, 75437, 75448, 75501, 75431, 75429, 75435, 75419, 75460, 75669, 75445, 75438, 75444, 75585, 75455, 75432, 75673, 75434, 75440, 75519, 75461, 75443, 75439, 75466, 75531, 75450, 75638, 75433, 75441, 75451, 75465, 75536, 75447, 75486, 75452, 75540, 75490, 75422, 75464, 75532. Wait queue: ."

```
Classic deadlock! 
My question is why would the maintenance function need an access exclsuive lock on the default partition when it is trying to create an new partition?

Also, we recently moved to a new schema using weekly partitions. We were previously using a daily partition strategy and we did not see these issues. We checked on cron job run details and did not find any failures on the daily partitioned DB.

Again, really appreciate your help in this regard. 🙇

**Reply by sriram-au** [reply]
Also, we don't have any data in the default.

**Reply by keithf4** [reply]
partman itself isn't taking any specific locks during routine maintenance, it's the underlying declarative postgresql commands. The default table does have a constraint on it and if you look at it you'll notice it's basically an anti-constraint of all the currently existing children. When a new child table is added, that constraint needs to be updated. Also if there is data in the default, it has to scan all of that data to ensure the new child table's constraint doesn't match any rows (it fails if it does). Not quite sure why it needs such a heavy lock to do so, especially if the default is empty. Would have to pose that question to the general or hackers mailing list.

My only suggestion for now would be to maybe try scheduling this one table's maintenance manually with the regular `run_maintenance()` function instead of the procedure. You could set this partition set's `automatic_maintenance` column in `part_config` to `off`, which stops that table's maintenance running when either maintenance routine is called without a parameter. Then pass this table in as a parameter with a direct call on its own schedule like this: `SELECT run_maintenance('public.command_record');`

Or if you don't really need the default you could drop it. Any attempt to insert data outside the tables covered by the children would then just result in an error.

Also, just an FYI for the upcoming 5.0 version, the weekly partition method changed a bit. Looking to have it out this fall, but then not sure how long it will take RDS to adopt it. https://github.com/pgpartman/pg_partman/blob/5.0.0-beta/doc/pg_partman_5.0.0_upgrade.md

**Comment by sriram-au** [comment]
>My only suggestion for now would be to maybe try scheduling this one table's maintenance manually with the regular run_maintenance() function instead of the procedure. You could set this partition set's automatic_maintenance column in part_config to off, which stops that table's maintenance running when either maintenance routine is called without a parameter. Then pass this table in as a parameter with a direct call on its own schedule like this: SELECT run_maintenance('public.command_record');

Alright, we have set this up in our staging env. However, we have not seen these deadlocks in in our lower environments, lets see what happens. The scheduled job is set to run tonight. Also, is there a difference between what is being run by the procedure and the function `run_maintenance`?

>partman itself isn't taking any specific locks during routine maintenance, it's the underlying declarative postgresql commands. The default table does have a constraint on it and if you look at it you'll notice it's basically an anti-constraint of all the currently existing children. 

This is an interesting point. We ran the alter table DDL (see below) in our production environment to attach the partition manually, which also created the anti-constraints on the default table. The command execution was almost instanteneous. If this was an issue with the underlying `declarative postgresql commands` ideally we should have seen the deadlock here as well. The read/write patterns on the DB was not very different from when we ran the attach partition manually and when the cron job ran last night.

```
CREATE TABLE command_record_p2023w38 PARTITION OF command_record FOR VALUES FROM ('2023-09-18 00:00:00+00') TO ('2023-09-25 00:00:00+00');
```

>Or if you don't really need the default you could drop it. Any attempt to insert data outside the tables covered by the children would then just result in an error.

Having a default table is useful when we may have to support data inserts outside the partition ranges. This paractically does not
 make sense for our use case, but still theoritically possible. Hence, we would like to keep the default table.

**Reply by keithf4** [reply, marked as answer]
The procedure just calls the function in a loop over all the partition sets. But the procedure runs individual transactions for each partition set vs one giant transaction if you called just the regular function. So it's mostly just a different transaction state. 

One other thing I thought of that you're likely not doing when you make a child table manually is an analyze on the partition set. In the past, the analyze was enabled by default, but later versions of pg_partman disabled it by default (https://github.com/pgpartman/pg_partman/blob/master/sql/functions/run_maintenance.sql#L146). It doesn't look like you're enabling it from what I can see, but it's something else to check. 

The deadlock really just seems to be a timing thing. Do you have any timings on how long the procedure call takes to run? Is it taking longer than it did in the past? partman's manitenance is never going to be as quick as simply adding a new child table since there are more actions taken inside the function. But that's also what using the procedure vs the function to do maintenance is supposed to help with, only leaving a transaction open long enough for a single set. But, calling a single partition set directly via the function should be a little faster as well.
